test_vm/src/lib.rs

```diff
@@ -0,0 +1,90 @@
+use cid::Cid;
+use fvm_ipld_hamt::{BytesKey, Hamt, Sha256};
+use fvm_shared::address::Address;
+use fvm_shared::bigint::bigint_ser;
+use fvm_shared::blockstore::MemoryBlockstore;
+use fvm_shared::econ::TokenAmount;
+use fvm_shared::encoding::tuple::*;
+use std::error::Error;
+use std::fmt;
+
+pub struct VM<'bs> {
+    store: &'bs MemoryBlockstore,
+    state_root: Cid,
+    actors_dirty: bool,
+    actors: Hamt<&'bs MemoryBlockstore, Actor, BytesKey>,
+}
+
+impl<'bs> VM<'bs> {
+    pub fn new(store: &'bs MemoryBlockstore) -> VM<'bs> {
+        let mut actors = Hamt::<&'bs MemoryBlockstore, Actor, BytesKey, Sha256>::new(store);
+        VM { store, state_root: actors.flush().unwrap(), actors_dirty: false, actors }
+    }
+
+    pub fn get_actor(&self, addr: Address) -> Result<Actor, TestVMError> {
+        match self.actors.get(&addr.to_bytes())? {
+            None => Err(vm_err("failed to get addr")),
+            Some(a) => Ok(a.clone()),
+        }
+    }
+
+    // blindly overwrite the actor at this address whether it previously existed or not
+    pub fn set_actor(&mut self, key: Address, a: Actor) -> Result<(), TestVMError> {
+        let _ = self.actors.set(key.to_bytes().into(), a)?;
+        Ok(())
+    }
+
+    pub fn checkpoint(&mut self) -> Result<Cid, TestVMError> {
+        self.state_root = self.actors.flush()?;
+        self.actors_dirty = false;
+        Ok(self.state_root)
+    }
+
+    pub fn rollback(&mut self, root: Cid) -> Result<(), TestVMError> {
```

Comment:
`Runtime ` is inconsistent, but I think `Cid` should be by reference too. Note how `load` takes a reference. It otherwise copies all the bytes, if I understand correctly.